x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/HasApplicationPrivilegesIT.java

```diff
@@ -0,0 +1,215 @@
+
+    @Before
+    public void setupClient() {
+        securityClient = new TestSecurityClient(adminClient());
+    }
+
+    public void testUserWithWildcardPrivileges() throws Exception {
+        var appName = randomApplicationName();
+
+        final Set<String> allApplications = new HashSet<>();
+        CheckedBiConsumer<String, Integer, IOException> createApplicationPrivilege = (app, index) -> {
+            createApplicationPrivilege(app, "priv_" + index, randomArray(1, 4, String[]::new, () -> randomActionName()));
+            allApplications.add(appName);
+        };
+
+        for (int i = randomIntBetween(0, 2); i > 0; i--) {
+            // Create 0 or more application privileges for this application
+            createApplicationPrivilege.accept(appName, i);
+        }
+
+        for (int i = randomIntBetween(0, 3); i > 0; i--) {
+            // Create 0 or more application privileges for other applications
+            createApplicationPrivilege.accept(randomValueOtherThan(appName, this::randomApplicationName), i);
+        }
+
+        // Define a role with all privileges (by wildcard) for this application
+        var roleName = randomAlphaOfLengthBetween(6, 10);
+        var singleAppOnly = randomBoolean();
+        createRole(roleName, singleAppOnly ? appName : "*", new String[] { "*" }, new String[] { "*" });
+
+        final Set<String> allRoles = new HashSet<>();
+        allRoles.add(roleName);
+
+        for (int i = randomIntBetween(0, 3); i > 0; i--) {
+            // Create 0 or more additional roles with privileges for one of the applications
+            var extraRoleName = randomValueOtherThanMany(allRoles::contains, () -> randomAlphaOfLengthBetween(8, 16));
+            createRole(
+                extraRoleName,
+                randomFrom(allApplications),
+                new String[] { "priv_" + i },
+                new String[] { "data/" + randomAlphaOfLength(6) }
+            );
+            allRoles.add(extraRoleName);
+        }
+
+        // Create a user with all (might be 1 or more) of the roles
+        var username = randomAlphaOfLengthBetween(8, 12);
+        var password = new SecureString(randomAlphaOfLength(12).toCharArray());
+        createUser(username, password, allRoles);
+
+        // Assert that has_privileges returns true for any arbitrary privilege or action in that application
+        var reqOptions = RequestOptions.DEFAULT.toBuilder()
+            .addHeader("Authorization", UsernamePasswordToken.basicAuthHeaderValue(username, password))
+            .build();
+
+        var testPrivilege = randomBoolean()
+            ? randomAlphaOfLengthBetween(4, 12)  // privilege name
+            : randomActionName() // action name
+        ;
+        var testResource = randomAlphaOfLengthBetween(4, 12);
+
+        {
+            final List<ResourcePrivileges> shouldHavePrivileges = hasPrivilege(
+                reqOptions,
+                appName,
```

Comment:
Can you elaborate?
My intent was that this `appName` is picked so that we know at the beginning of the test which app we're going to use for `_has_privileges` and then randomise everything around it. 

- whether or not that app has defined privileges is randomised (first `for` loop).
- whether or not _other_ apps have defined privileges is randomised (second `for` loop).
- whether the superuser-style role has access to _all apps_ or just _this app_ is randomised.

I don't think randomising here would introduce anything new. I can do it (assuming `singleAppOnly == false`), but does it increase the scope of the testing?

Reply:
You are right. Somehow I thought this block was only for `singleAppOnly == false`. But it is not, so all good.